src/topology/metric_space/dilation.lean

```diff
@@ -0,0 +1,427 @@
+
+instance : has_coe_to_fun (dilation α β) (λ _, α → β) := fun_like.has_coe_to_fun
+
+@[simp] lemma to_fun_eq_coe {f : dilation α β} : f.to_fun = (f : α → β) := rfl
+
+@[simp] lemma coe_mk (f : α → β) (h) : ⇑(⟨f, h⟩ : dilation α β) = f := rfl
+
+lemma congr_fun {f g : dilation α β} (h : f = g) (x : α) : f x = g x := fun_like.congr_fun h x
+lemma congr_arg (f : dilation α β) {x y : α} (h : x = y) : f x = f y := fun_like.congr_arg f h
+
+@[ext] theorem ext {f g : dilation α β} (h : ∀ x, f x = g x) : f = g :=
+fun_like.ext f g h
+
+lemma ext_iff {f g : dilation α β} : f = g ↔ ∀ x, f x = g x := fun_like.ext_iff
+
+@[simp] lemma mk_coe (f : dilation α β) (h) : dilation.mk f h = f := ext $ λ _, rfl
+
+/-- Copy of a `dilation` with a new `to_fun` equal to the old one. Useful to fix definitional
+equalities. -/
+protected def copy (f : dilation α β) (f' : α → β) (h : f' = ⇑f) : dilation α β :=
+{ to_fun := f',
+  edist_eq' := h.symm ▸ f.edist_eq' }
+
+/-- The ratio of a dilation `f`. Uses `Exists.some`, the `some_spec` counterpart
+is the lemma `edist_eq` below -/
+def ratio [dilation_class F α β] (f : F) : ℝ≥0 :=
+(dilation_class.edist_eq' f).some
+
+@[simp] lemma edist_eq [dilation_class F α β] (f : F) (x y : α) :
+  edist (f x) (f y) = ratio f * edist x y :=
+(dilation_class.edist_eq' f).some_spec x y
+
+@[simp] lemma dist_eq {α β} {F : Type*}
+  [pseudo_metric_space α] [pseudo_metric_space β] [dilation_class F α β]
+  (f : F) (x y : α) : dist (f x) (f y) = ratio f * dist x y :=
+by simp only [dist_edist, ennreal.to_real_mul, edist_eq, ennreal.coe_to_real]
+
+@[simp] lemma nndist_eq {α β} {F : Type*}
+  [pseudo_metric_space α] [pseudo_metric_space β] [dilation_class F α β]
+  (f : F) (x y : α) : nndist (f x) (f y) = ratio f * nndist x y :=
+begin
+  apply subtype.ext _,
+  simp only [coe_nndist, dist_eq, nonneg.coe_mul],
+end
+
+/-- The `ratio` is equal to the distance ratio for any two points with nonzero finite distance.
+`dist` and `nndist` versions below -/
+lemma ratio_unique [dilation_class F α β] {f : F} {x y : α} {r : ℝ≥0}
+  (hxy : edist x y ≠ 0 ∧ edist x y ≠ ⊤) (hr : edist (f x) (f y) = r * edist x y) :
+  r = ratio f :=
+begin
+  have h := edist_eq f x y,
+  rwa [hr, ennreal.mul_eq_mul_right hxy.1 hxy.2, ennreal.coe_eq_coe] at h,
+end
+
+/-- The `ratio` is equal to the distance ratio for any two points
+with nonzero finite distance; `dist version` -/
+lemma ratio_unique' {α β} {F : Type*} [pseudo_metric_space α] [pseudo_metric_space β]
+  [dilation_class F α β] {f : F} {x y : α} {r : ℝ≥0}
+  (hxy : dist x y ≠ 0) (hr : dist (f x) (f y) = r * dist x y) :
+  r = ratio f :=
+begin
+  have h := dist_eq f x y,
+  rw [hr, (mul_eq_mul_right_iff)] at h,
+  have := or.resolve_right h hxy,
+  rwa nnreal.coe_eq at this,
+end
+
+/-- The `ratio` is equal to the distance ratio for any two points
+with nonzero finite distance; `nndist version` -/
+lemma ratio_unique'' {α β} {F : Type*} [pseudo_metric_space α] [pseudo_metric_space β]
```

Comment:
Using `'` and `''` on the names of lemmas to distinguish different versions isn't very friendly to users; names such as `ratio_unique_of_dist_ne_zero` would be friendlier.